tensorflow_model_optimization/python/core/common/keras/compression/internal/optimize.py

```diff
@@ -264,7 +264,8 @@ def _map_to_training_weights(
           # suggests that the `value` cannot be any arbitrary shape and
           # only a single scalar value. It works in this implementation
           # to make `value` any tensor - check this.
-          training_weights.append(weight_repr.initializer(shape=None))
+          training_weights.append(weight_repr.initializer(
+              shape=None, dtype=weight_repr.dtype))
```

Comment:
Trying to understand this:

[tf.keras.initializers.Constant](https://www.tensorflow.org/api_docs/python/tf/keras/initializers/Constant )suggests that the default is float32. Is that what you observed before this change? 

And now for the weight clustering example, I guess the AbstractClusteringAlgorithm API docs suggests that
a) pulling_indices has type int32, so now it'll be that as expected
b) clustering_centroids has the same dtype as the original weights, which for [Conv2D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Conv2D) seems to get its dtype from [tf.keras.layers.Layer](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Layer) which seems to be float32 by default also.

So going from float32 -> (float 32 or int32) improved the compression ratio from applying gzip? Or is something else happening.

Reply:
Yes that's almost correct.

> pulling_indices dtype is int64, but only use few integer (0~# of cluster-1) so gzip compress significantly decrease model size. (but shape is as same as original weights.)

Reply:
Got it, so this is mainly beneficial not for the centroids but for the pulling indices.

Reply:
Beyond this PR, we should try to explore some implementation so that the user doesn't have to add this boilerplate information.

One option is to keep WeightRepr, but remove the shape, dtype, and initializer parameters so that it takes a tensor instead. Under the hood, we have similar code that converts that tensor into those three fields. The assumption is that a user wouldn't need to directly set those parameters.